File: 백준/Gold/12100. 2048 （Easy）/2048 （Easy）.cc
#include <bits/stdc++.h>
using namespace std;
#define fastio cin.tie(0)->sync_with_stdio(0)
#define ll long long
#define X first
#define Y second
#define endl "\n"

int n;
int br1[22][22];
int br2[22][22];

void rotate() {
  int tmp[22][22];
  // copy br2
  for(int i=0;i<n;++i){
    for(int j=0;j<n;++j){
      tmp[i][j] = br2[i][j];
    }
  }
  
  for(int i=0;i<n;++i){
    for(int j=0;j<n;++j){
      br2[i][j] = tmp[n-1-j][i];
    }
  }
}

void tilt(int dir) {
  while(dir--) rotate();
  for(int i=0;i<n;i++) {
    int tilted[22] = {};
    int idx = 0;
    for(int j=0;j<n;++j){
      if(br2[i][j]==0) continue;
      if(tilted[idx]==0) {
        tilted[idx] = br2[i][j];
      }
      else if(tilted[idx] == br2[i][j]) {
        tilted[idx++] *= 2;
      }
      else {
        tilted[++idx] = br2[i][j];
      }
    }
    
    for(int j=0;j<n;++j) br2[i][j] = tilted[j];
  }
}

int main(void) {
  fastio;
  cin>>n;
  //init board
  for(int i=0;i<n;++i){
    for(int j=0;j<n;++j){
      cin>>br1[i][j];
    }
  }
  
  int mx = 0;
  for(int tmp=0; tmp<1024; ++tmp) {
    // copy br1 to br2
    for(int i=0;i<n;++i){
      for(int j=0;j<n;++j) {
        br2[i][j] = br1[i][j];
      }
    }
    
    int brute = tmp;
    for(int i=0;i<5;++i) {
      int dir = brute%4;
      brute /= 4;
      tilt(dir);
    }
    
    for(int i=0;i<n;++i){
      for(int j=0;j<n;++j){
        mx = max(mx, br2[i][j]);
      }
    }
  }
  cout<<mx;
  return 0;
}

File: 백준/Gold/12100. 2048 （Easy）/2048 （Easy）.py
import sys
fi = sys.stdin.readline
sys.setrecursionlimit(10**8)
    
def solution():
    global br2, n
    n = int(fi().rstrip())
    br1 = [list(map(int, fi().split())) for _ in range(n)]
    
    mx = 0
    for tmp in range(1<<(2*5)):
        br2 = br1[:]
        brute = tmp
        for i in range(5):
            dir = brute % 4
            brute = brute // 4
            tilt(dir)
        
        for i in range(n):
            for j in range(n):
                mx = max(mx, br2[i][j])
    print(mx)
    
def rotate():
    global br2
    tmp = list(map(list, zip(*br2[::-1])))
    br2 = tmp[:]

def tilt(dir):
    global br2, n
    while dir > 0:
        # 방향 만큼 돌려주기
        rotate()
        dir -= 1
    
    for i in range(n):
        # 돌리고 난 후 합칠 배열 선언 한 행씩 계산
        tilted = [0] * n
        idx = 0
        for j in range(n):
            # 해당 칸에 숫자가 없다면 스킵
            if br2[i][j] == 0: continue
            # 돌린 후 배열에 해당 인덱스 위치가 비어 있다면 숫자 넣어주기
            if tilted[idx] == 0:
                tilted[idx] = br2[i][j]
            # 돌린 후 배열 숫자와 돌리기 전 배열 숫자가 같다면
            elif tilted[idx] == br2[i][j]:
                # 두배로 만들어 주고(합쳤으니까)
                tilted[idx] *= 2
                # 인덱스 늘려주기
                idx += 1
            # 합칠 수 없는 숫자라면(이미 합친 숫자이거나 다른 숫자) 그 다음 칸에 숫자 넣어주기
            else:
                idx += 1
                tilted[idx] = br2[i][j]
        # 합친 배열로 해당 행 교체해주기
        br2[i] = tilted[:]
            
if __name__ == '__main__':
    solution()